spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java

```diff
@@ -0,0 +1,381 @@
+              toColumnName(leftChild(predicate)),
+              convertUTF8StringIfNecessary(rightChild(predicate)).toString());
+      }
+    }
+
+    return null;
+  }
+
+  private static Object convertUTF8StringIfNecessary(Literal literal) {
+    if (literal.value() instanceof UTF8String) {
+      return ((UTF8String) literal.value()).toString();
+    }
+    return literal.value();
+  }
+
+  private static Expression handleEqual(String attribute, Object value) {
+    if (NaNUtil.isNaN(value)) {
+      return isNaN(attribute);
+    } else {
+      return equal(attribute, value);
+    }
+  }
+
+  private static boolean hasNoInFilter(Predicate predicate) {
+    Operation op = FILTERS.get(predicate.name());
+
+    if (op != null) {
+      switch (op) {
+        case AND:
+          And andPredicate = (And) predicate;
+          return hasNoInFilter(andPredicate.left()) && hasNoInFilter(andPredicate.right());
+        case OR:
+          Or orPredicate = (Or) predicate;
+          return hasNoInFilter(orPredicate.left()) && hasNoInFilter(orPredicate.right());
+        case NOT:
+          Not notPredicate = (Not) predicate;
+          return hasNoInFilter(notPredicate.child());
+        case IN:
+          return false;
+        default:
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  private static boolean valid(Predicate predicate) {
+    Operation op = FILTERS.get(predicate.name());
+    if (op != null) {
+      switch (op) {
+        case IS_NULL:
+        case NOT_NULL:
+          return child(predicate) instanceof NamedReference;
+
+        case LT:
+        case LT_EQ:
+        case GT:
+        case GT_EQ:
+        case EQ:
+          if (predicate.children().length != 2) {
+            return false;
+          }
+          return namedRefCompLit(predicate) || litCompNamedRef(predicate);
+
+        case IN:
+          if (!(leftChild(predicate) instanceof NamedReference)) {
+            return false;
+          } else {
+            return Arrays.stream(predicate.children())
+                .skip(1)
+                .allMatch(val -> val instanceof Literal);
+          }
+
+        case NOT:
+          Not notFilter = (Not) predicate;
+          return valid(notFilter.child());
+
+        case AND:
+          And andFilter = (And) predicate;
+          return valid(andFilter.left()) && valid(andFilter.right());
+
+        case OR:
+          Or orFilter = (Or) predicate;
+          return valid(orFilter.left()) && valid(orFilter.right());
+
+        case STARTS_WITH:
+          if (predicate.children().length != 2) {
+            return false;
+          }
+          return namedRefCompLit(predicate)
+              && ((Literal<?>) rightChild(predicate)).value() instanceof String;
+
+        case TRUE:
+        case FALSE:
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static boolean namedRefCompLit(Predicate predicate) {
```

Comment:
I think splitting this into `isRef` and `isLiteral` should allow us to combine the validation and conversion into one pass. Then we won't have to jump between methods when looking at the code.

Reply:
Agree!